### PR TITLE
Feature: unified Over-The-Air (OTA) upgrades

### PR DESCRIPTION
**Pull request by slaff**
This feature adds architecture-agnostic OTA support. Some of the highlights here:

- Added component `Ota` that encapsulates the most important Ota functions and works with storage partitions.
- Added component `OtaNetwork` to add network support for the Ota functions.
- Sample Basic_rBoot is replaced with Basic_Ota. Direct use or rboot component inside your application is discouraged.
- HttpServer_FirmwareUpdate is updated to use the new Ota component.
- Sample OtaUpgradeMqtt::Upgrade is updated to use new Ota component.

> With a bit of work the OTA component could support also ESP32.

Related to #2328

### CODE REVIEW

File: Sming/Arch/Esp32/spiffs-two-roms.hw [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{
+	"name": "Two ROM slots with single SPIFFS",
+	"base_config": "spiffs",
+	"partitions": {
+	    "factory": {
+	       "address": "0x10000",
```
Comment (mikee47) [inline Sming/Arch/Esp32/spiffs-two-roms.hw:6]: Only

File: Sming/Libraries/OtaUpgradeMqtt/samples/Upgrade/app/application.cpp [hunk 2]
```diff
@@ -113,15 +97,20 @@ void showInfo()
...
-	debug_d("Count: %d", conf.count);
-	debug_d("ROM 0: 0x%08x", conf.roms[0]);
-	debug_d("ROM 1: 0x%08x", conf.roms[1]);
-	debug_d("ROM 2: 0x%08x", conf.roms[2]);
-	debug_d("GPIO ROM: %d", conf.gpio_rom);
+	auto part = OtaManager.getRunningPartition();
 
-	Serial.printf(_F("\r\nCurrently running rom %d. Application version: %s\r\n"), conf.current_rom, APP_VERSION);
+	Serial.printf(_F("\r\nCurrently running %s: 0x%08x. Application version: %s\r\n"), part.name(), part.address(),
+				  APP_VERSION);
```
Comment (mikee47) [inline Sming/Libraries/OtaUpgradeMqtt/samples/Upgrade/app/application.cpp:113]: `part.name.c_str()`

**Review by mikee47 — COMMENTED** [review]
Much, much better!

**Comment by slaff** [comment]
@mikee47 The OtaUpgrade library uses the RBOOT_xxx variables to read their location and address. But in ESP32 they are not present. 
```bash
	$(Q) $(OTATOOL) mkfile \
		$(OTA_CRYPTO_FEATURES_IMAGE) \
		$(if $(OTA_CRYPTO_FEATURES_IMAGE),--key=$(OTA_KEY_IMAGE)) \
		--rom=$(RBOOT_ROM_0_BIN)@$(RBOOT_ROM0_ADDR) \
		$(if $(RBOOT_ROM_1_BIN),--rom=$(RBOOT_ROM_1_BIN)@$(RBOOT_ROM1_ADDR)) \
```
Is there an easy to get them from the hardware configuration using one of the python scripts that come with the partitioning?  Can we somehow query the JSON hardware configuration without having to install aditional python library as https://github.com/h2non/jsonpath-ng for example?

**Comment by mikee47** [comment]
@slaff Yes, there is an easy way! If you look in `out\Esp32\debug\hwconfig.mk` there are corresponding definitions which can be used without resorting to JSON. For example:

`RBOOT_ROM0_BIN` -> `PARTITION_rom0_FILENAME`

**Pull request merged**